src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp

```diff
@@ -76,29 +87,147 @@ ShenandoahHeuristics::~ShenandoahHeuristics() {
   FREE_C_HEAP_ARRAY(RegionGarbage, _region_data);
 }
 
-size_t ShenandoahHeuristics::select_aged_regions(size_t old_available, size_t num_regions, bool* preselected_regions) {
+typedef struct {
+  ShenandoahHeapRegion* _region;
+  size_t _live_data;
+} AgedRegionData;
+
+static int compare_by_aged_live(AgedRegionData a, AgedRegionData b) {
+  if (a._live_data < b._live_data)
+    return -1;
+  else if (a._live_data > b._live_data)
+    return 1;
+  else return 0;
+}
+
+// Returns bytes of old-gen memory consumed by selected aged regions
+size_t ShenandoahHeuristics::select_aged_regions(size_t old_available, size_t num_regions, bool preselected_regions[]) {
   ShenandoahHeap* heap = ShenandoahHeap::heap();
   assert(heap->mode()->is_generational(), "Only in generational mode");
-
+  ShenandoahMarkingContext* const ctx = heap->marking_context();
   size_t old_consumed = 0;
+  size_t promo_potential = 0;
+  size_t anticipated_promote_in_place_live = 0;
+
+  heap->clear_promotion_in_place_potential();
+  heap->clear_promotion_potential();
+  size_t candidates = 0;
+  size_t candidates_live = 0;
+  size_t old_garbage_threshold = (ShenandoahHeapRegion::region_size_bytes() * ShenandoahOldGarbageThreshold) / 100;
+  size_t promote_in_place_regions = 0;
+  size_t promote_in_place_live = 0;
+  size_t promote_in_place_pad = 0;
+  size_t anticipated_candidates = 0;
+  size_t anticipated_promote_in_place_regions = 0;
+
+  // Sort the promotion-eligible regions according to live-data-bytes so that we can first reclaim regions that require
+  // less evacuation effort.  This prioritizes garbage first, expanding the allocation pool before we begin the work of
+  // reclaiming regions that require more effort.
+  AgedRegionData* sorted_regions = (AgedRegionData*) alloca(num_regions * sizeof(AgedRegionData));
   for (size_t i = 0; i < num_regions; i++) {
-    ShenandoahHeapRegion* region = heap->get_region(i);
-    if (in_generation(region) && !region->is_empty() && region->is_regular() && (region->age() >= InitialTenuringThreshold)) {
-      size_t promotion_need = (size_t) (region->get_live_data_bytes() * ShenandoahEvacWaste);
-      if (old_consumed + promotion_need < old_available) {
+    ShenandoahHeapRegion* r = heap->get_region(i);
+    if (r->is_empty() || !r->has_live() || !r->is_young() || !r->is_regular()) {
+      continue;
+    }
+    if (r->age() >= InitialTenuringThreshold) {
+      r->save_top_before_promote();
```

Comment:
We have to save top before promote, even if we do not  promote in place, because evacuator asks "what was garbage before top was saved"?  Need to think about tis...

Reply:
the evacuator asks what was top at mark end?